---
title: "MIGEC ANALYSIS REPORT"
output:
  html_document:
    highlight: tango
    theme: united
    toc: yes    
    toc_depth: 3
---

> Working directory: `r format(getwd())`
>
> Created: `r format(Sys.time(), '%d %B, %Y')`

---

```{r, echo=FALSE, message=FALSE, warning=FALSE}
# Modify these fields to set directories
# If some analysis were not performed, set corresponding values to NULL
checkout_path  = "checkout/"
histogram_path = "histogram/"
assemble_path  = "assemble/"
cdrblast_path  = "cdrblast/"
cdrfinal_path  = "cdrfinal/"
```

## Step I: Checkout

This section summarizes the efficiency of sample barcode matching and unique molecular identifier (UMI) extraction.

Command line:


```{r, echo=FALSE} 
message(ifelse(!(is.null(checkout_path)), readLines(paste(checkout_path, "/checkout.cmd.txt", sep = "/")), "none"))
```

```{r, echo=FALSE, message=FALSE, warning=FALSE}
require(ggplot2); require(reshape);

# load and groom data
if (!is.null(checkout_path)) {
df <- read.table(paste(checkout_path, "/checkout.log.txt", sep = "/"), header=T, comment ="", row.names = NULL)
df <- df[,3:6]
df[,2:4] <- apply(df[, 2:4], 2, as.numeric)
df[,1] <- as.character(df[, 1])
df <- replace(df, is.na(df), 0)
colnames(df) <- c("sample", "master", "slave", "overlapped")

# summary
df <- aggregate(cbind(master, slave, overlapped) ~ sample, data=df, FUN=sum)

#print(df)

total <- sum(df[,3])
found <- df[1,]
found[,2:4]<-colSums(df[,2:4])
missing <- df[df$sample=='undef-m',]
found[,2:4] <- found[,2:4] - missing[,2:4]
missing[,3] <- df[df$sample=='undef-s',][,3]
found[,3] <- found[,3] - missing[,3]

df.s <-rbind(found, missing)
df.s[,1]<-c('found','missing')
df.s <- melt(df.s)

df <- df[- grep("undef", df$sample),]
df <- melt(df)
}
```

### Fig.1 Total extraction efficiency

Overall extraction efficiency, showing the success rate of finding the 

* **master**, main barcode containing sample-specific sequence

* **slave**, auxiallary barcode, used to protect from contamination, only checked if **master** is detected 

and the rate of read overlapping, **overlapped** (if enabled).

```{r, echo=FALSE, message=FALSE}
percent <- function(x) {
  paste(as.character(round(x*1000)/10),"%",sep="")
}
p<-"Nothing to plot"
if (!is.null(checkout_path)) {
p<-ggplot(df.s, aes(x=sample,y=value,fill=variable)) + 
  geom_bar(stat='identity', position='dodge', size = 0.2, colour = "gray30") +
  scale_y_continuous(expand = c(0, 0), limits=c(0,max(df.s$value)*1.2)) + xlab("") + ylab("reads") +
  scale_fill_brewer(palette="Paired") +  
  geom_text(position = position_dodge(width=1), angle=90, size = 3, hjust = -0.1, aes(label=value)) +
  geom_text(position = position_dodge(width=1), vjust=-1, size = 3, aes(y=0,label=percent(value/total))) +
  theme_bw()
remove(df.s)
}
p
```

### Fig.2 Extraction efficiency by sample

Distribution of reads recovered by sample

```{r, echo=FALSE, message=FALSE}
p<-"Nothing to plot"
if (!is.null(checkout_path)) {
p<-ggplot(df, aes(x=sample,y=value,fill=variable)) + 
  geom_bar(stat='identity', position='dodge', size = 0.2, colour = "gray30") + 
  geom_text(position = position_dodge(width=1), size = 3, hjust = -0.1, aes(label=value)) +
  coord_flip() +
  scale_y_log10(expand = c(0, 0), limits=c(1,max(df$value)*10)) + ylab("reads") + xlab("") +
  scale_fill_brewer(palette="Paired") +
  theme_bw()
remove(df)
}
p
```

---

## Step II: Histogram

This section contains data on sequencing depth based on counting the number of reads tagged with the same UMI, i.e. reads that belong to the same Molecular Identifier Group (MIG).

Command line: 

```{r, echo=FALSE} 
message(ifelse(!(is.null(histogram_path)), readLines(paste(histogram_path, "/histogram.cmd.txt", sep = "/")), "none"))
```

### Fig.3 UMI coverage plot

Below is the plot of MIG size distribution that was used to estimate MIG size threshold for assembly (shown by dashed lines).

> Note that the distribution is itself weighted by the number of reads to highlight the mean coverage value and to reflect the percent of reads that will be retained for a given MIG size threshold.

```{r, echo=FALSE, message=FALSE, warning=FALSE}
require(plyr); require(RColorBrewer)

p<-"Nothing to plot"
if (!(is.null(histogram_path))) {
df <- read.table(paste(histogram_path, "/overseq.txt", sep = "/"), header=T, comment ="")

df.a <- read.table(paste(histogram_path, "/estimates.txt", sep = "/"), header=T, comment ="")
df.a <- aggregate(OVERSEQ_THRESHOLD ~ X.SAMPLE_ID, data=df.a, FUN=max)
df <- merge(df, df.a, by = "X.SAMPLE_ID")
df$OVERSEQ_THRESHOLD <- as.character(df$OVERSEQ_THRESHOLD)

# compute peak values
df$peak <- numeric(nrow(df))

for (i in 1:nrow(df)) {
  df$peak[i] <- which.max(df[i,3:ncol(df)])
}

df$peak <- as.character(df$peak)

# groom data
df <- melt(df, ids = c(X.SAMPLE_ID, SAMPLE_TYPE, peak, OVERSEQ_THRESHOLD))
df$peak <- as.numeric(df$peak)
df$variable <- as.character(df$variable)
df$variable <- as.numeric(gsub('X', '', df$variable))
df$OVERSEQ_THRESHOLD <- as.numeric(df$OVERSEQ_THRESHOLD)
colnames(df) <- c("sample", "sample.type", "threshold", "peak", "mig.size", "count")

# summarize by sample type, normalize within sample
df <- aggregate(count ~ sample + mig.size + threshold + peak, data=df, FUN=sum)
df.n <- ddply(df,.(sample),transform,count=count/sum(count))
df.n$sample <-factor(df.n$sample, levels=unique(df.n[order(df.n$peak), "sample"]))

# plotting

pal <- colorRampPalette(brewer.pal(9, "GnBu"))(length(unique(df$sample)))

p <- ggplot(df.n, aes(x=mig.size, color = sample, weight=count)) + 
  geom_vline(aes(xintercept=threshold, color = sample), linetype = "longdash") +
  stat_density(position='identity', adjust = 0.5, size = 0.5, fill = NA) +
  scale_color_manual(values = pal) +
  scale_fill_manual(values = pal) +
  scale_y_continuous(expand=c(0,0)) +
  scale_x_log10(breaks=c(2^(0:7), 512, 2048, 8192, 32768), limits=c(1, 32768), expand=c(0,0)) + 
  ylab("frequency") + xlab("MIG size, reads") + theme_bw() +
  theme(legend.direction ="horizontal",legend.position = "bottom") +
  guides(color=guide_legend(ncol=4))
remove(df)
remove(df.n)
}
p
```

---

## Step III: Assemble

This section contains information on the assembly of MIG consensuses from raw reads.

Command line:

```{r, echo=FALSE} 
message(ifelse(!(is.null(assemble_path)), readLines(paste(assemble_path, "/assemble.cmd.txt", sep = "/")), "none"))
```


```{r, echo=FALSE, message=FALSE, warning=FALSE}
require(scales)

if (!is.null(assemble_path)) {
df <- read.table(paste(assemble_path, "/assemble.log.txt", sep = "/"), header=T, comment ="")
df$READS_DROPPED_WITHIN_MIG = df$READS_DROPPED_WITHIN_MIG_1 + df$READS_DROPPED_WITHIN_MIG_2
df <- data.frame(sample <- df$X.SAMPLE_ID, 
                 migs.assembled <- df$MIGS_GOOD_TOTAL,
                 umi.fraction.assembled <- df$MIGS_GOOD_TOTAL / df$MIGS_TOTAL,
                 reads.fraction.assembled <- df$READS_GOOD_TOTAL / df$READS_TOTAL,
                 reads.dropped.within.mig <- df$READS_DROPPED_WITHIN_MIG / df$READS_TOTAL)
df <- aggregate(cbind(migs.assembled, umi.fraction.assembled, reads.fraction.assembled, reads.dropped.within.mig) ~ sample, data=df, FUN=sum)
df$sample <-factor(df$sample, levels=df[order(df$migs.assembled), "sample"])
df <- melt(df)

pal <- colorRampPalette(brewer.pal(9, "GnBu"))(length(unique(df$sample)))

plotAsm.1 <- function(dd) {
  ggplot(dd, aes(x=sample, fill = sample, y=value)) +
    geom_bar(stat='identity', size = 0.2, colour = "gray30") + 
    geom_text(position = position_dodge(width=1), size = 3, hjust = -0.1, aes(label=value)) +
    coord_flip() +    
    scale_fill_manual(values = pal) + guides(fill = F) +
    xlab("") + ylab("") +
    theme_bw()
}
plotAsm.2 <- function(dd) {
  ggplot(dd, aes(x=sample, fill = sample, y=value)) +
    geom_bar(stat='identity', size = 0.2, colour = "gray30") + 
    geom_text(position = position_dodge(width=1), size = 3, hjust = -0.1, aes(label=round(value*1000)/10)) +
    coord_flip() +    
    scale_fill_manual(values = pal) + guides(fill = F) +
    xlab("") + ylab("") +
    theme_bw()
}
}
```

### Fig.4 Number of assembled MIGs

Below is a plot showing the total number of assembled MIGs per sample. The number of MIGs should be interpreted as the total number of starting molecules that have been successfully recovered.

```{r, echo=FALSE, message=FALSE, warning=FALSE}
p<-"Nothing to plot"
if (!is.null(assemble_path)) {
df.1 <- subset(df, variable == "migs.assembled")
p<-plotAsm.1(df.1) + scale_y_continuous(expand = c(0, 0), limits=c(0,max(df.1$value)*1.2))
}
p
```

### Fig.5 Frequency of productive UMI tags

Next comes the plot showing the fraction of UMIs that have resulted in assembled consensuses.

> This value is typically low, as it accounts for UMIs filtered due to not passing MIG size threshold. Such UMIs are mostly sequencing errors which generate a high amount of artificial UMI diversity.

```{r, echo=FALSE, message=FALSE}
p<-"Nothing to plot"
if (!is.null(assemble_path)) {
df.1 <- subset(df, variable == "umi.fraction.assembled")
p<-plotAsm.2(df.1) + scale_y_continuous(labels = percent_format(), expand = c(0, 0), limits=c(0,max(df.1$value)*1.2))
}
p
```

### Fig.6 Fraction of assembled reads

The fraction of reads contained within assembled MIGs when compared to total number of reads. This number should be high for a high-quality experiment.

```{r, echo=FALSE, message=FALSE}
p<-"Nothing to plot"
if (!is.null(assemble_path)) {
df.1 <- subset(df, variable == "reads.fraction.assembled")
p<-plotAsm.2(df.1) + scale_y_continuous(labels = percent_format(), expand = c(0, 0), limits=c(0,max(df.1$value)*1.2))
}
p
```

### Fig.7 Fraction of reads not fitting MIG consensus

The fraction of reads that were dropped during consensus assembly. Reads are being dropped if they are found to be substantially different from reads that form the core of consensus sequence. High numbers indicate low sequencing quality and/or presence of library preparation artifacts.

```{r, echo=FALSE, message=FALSE}
p<-"Nothing to plot"
if (!is.null(assemble_path)) {
df.1 <- subset(df, variable == "reads.dropped.within.mig")
p<-plotAsm.2(df.1) + scale_y_continuous(labels = percent_format(), expand = c(0, 0), limits=c(0,max(df.1$value)*1.2))
remove(df.1)
remove(df)
}
p
```

---

## Step IV: CdrBlast

This section contains results of the V(D)J segment mapping and CDR3 extraction algorithm running both for raw and assembled reads.

Command line:

```{r, echo=FALSE} 
message(ifelse(!(is.null(cdrblast_path)), readLines(paste(cdrblast_path, "/cdrblast.cmd.txt", sep = "/")), "none"))
```


```{r, echo=FALSE, message=FALSE, warning=FALSE}
if (!is.null(cdrblast_path)) {
df <- read.table(paste(cdrblast_path, "/cdrblast.log.txt", sep = "/"), header=T, comment ="")

df.m <- data.frame(sample = df$X.SAMPLE_ID, type = df$DATA_TYPE,
                 final.count = df$EVENTS_GOOD, 
                 qual.rate = df$EVENTS_GOOD / df$EVENTS_MAPPED,                 
                 map.rate = df$EVENTS_MAPPED / df$EVENTS_TOTAL)

df.m$metric <- as.factor("mig")
df.m <- df.m[- grep("raw", df.m$type),]
lvl <- df.m[order(df.m$final.count), "sample"]

df.r <- data.frame(sample = df$X.SAMPLE_ID, type = df$DATA_TYPE,
                 final.count = df$READS_GOOD, 
                 qual.rate = df$READS_GOOD / df$READS_MAPPED,
                 map.rate = df$READS_MAPPED / df$READS_TOTAL)
df.r$metric <- as.factor("read")

df <- rbind(df.m, df.r)
df <- melt(df)
df$sample <-factor(df$sample, levels=lvl)
pal <- colorRampPalette(brewer.pal(9, "GnBu"))(length(unique(df$sample)))

plotCdr.1 <- function(dd) {
  ggplot(dd, aes(x=sample, fill = sample, y=value)) +
    geom_bar(stat='identity', size = 0.2, colour = "gray30") + 
    geom_text(position = position_dodge(width=1), size = 3, hjust = -0.1, aes(label=value)) +
    coord_flip() +
    scale_fill_manual(values = pal) + guides(fill = F) +
    xlab("") + ylab("") + 
    theme_bw()
}

plotCdr.2 <- function(dd) {
  ggplot(dd, aes(x=sample, fill = sample, y=value)) +
    geom_bar(stat='identity', size = 0.2, colour = "gray30") + 
    geom_text(position = position_dodge(width=1), size = 3, hjust = -0.1, aes(label=round(value*1000)/10)) +
    coord_flip() +
    scale_fill_manual(values = pal) + guides(fill = F) +
    xlab("") + ylab("") + 
    theme_bw()
}
}
```

### Fig.8 Number of CDR3-containing MIGs

The plot below shows the total number of MIGs that contain good-quality CDR3 region in the consensus sequence

```{r, echo=FALSE, message=FALSE, warning=FALSE}
p <- "Nothing to plot"
if (!is.null(cdrblast_path)) {
df.s <- subset(df, variable == "final.count" & type == "asm" & metric == "mig")
p<-plotCdr.1(df.s) + scale_y_continuous(expand = c(0, 0), limits=c(0,max(df.s$value)*1.2))
}
p
```

### Fig.9 Number of CDR3-containing reads

Total number of reads that contain good-quality CDR3 region in raw reads

```{r, echo=FALSE, message=FALSE, warning=FALSE}
p <- "Nothing to plot"
if (!is.null(cdrblast_path)) {
df.s <- subset(df, variable == "final.count" & type == "asm" & metric == "read")
p<-plotCdr.1(df.s) + scale_y_continuous(expand = c(0, 0), limits=c(0,max(df.s$value)*1.2))
}
p
```

### Fig.10 CDR3 extraction rate

Mapping rate, the fraction of reads/MIGs that contain a CDR3 region

> Panels show assembled (**asm**) and unprocessed (**raw**) data. Values are given in number of molecules (**mig**, assembled samples only) and the corresponding read count (**read**)

```{r, echo=FALSE, message=FALSE, warning=FALSE}
p <- "Nothing to plot"
if (!is.null(cdrblast_path)) {
df.s <- subset(df, variable == "map.rate")
p<-plotCdr.2(df.s) + scale_y_continuous(labels = percent_format(), expand = c(0, 0), limits=c(0,max(df.s$value)*1.2)) + facet_grid(.~type+metric)
}
p
```

### Fig.11 Low-quality CDR3 filtering rate

Good-quality CDR3 sequence rate, the fraction of CDR3-containing reads/MIGs that pass quality filter

> Note that while raw data is being filtered based on Phred quality score, consensus quality score (CQS, the ratio of major variant) is used for assembled data

```{r, echo=FALSE, message=FALSE, warning=FALSE}
p <- "Nothing to plot"
if (!is.null(cdrblast_path)) {
df.s <- subset(df, variable == "qual.rate")
p<-plotCdr.2(df.s) + scale_y_continuous(labels = percent_format(), expand = c(0, 0), limits=c(0,max(df.s$value)*1.2)) + facet_grid(.~type+metric)
remove(df.s)
remove(df.r)
remove(df.m)
remove(df)
}
p
```

---

## Step V: CdrBlastFitler

This section contains the result of hot-spot error filtering stage, non-functional clonotype filtering (**if enabled**) and final statistics.

Command line:

```{r, echo=FALSE} 
message(ifelse(!(is.null(cdrfinal_path)), readLines(paste(cdrfinal_path, "/cdrblastfilter.cmd.txt", sep = "/")), "none"))
```


```{r, echo=FALSE, message=FALSE, warning=FALSE}
if (!is.null(cdrfinal_path)) {
nc_filter_enabled <- file.exists(paste(cdrfinal_path, "/cdrblastfilter.nc.txt", sep = "/"))
df <- read.table(paste(cdrfinal_path, "/cdrblastfilter.log.txt", sep = "/"), header=T, comment ="")
df <- data.frame(sample = df$X.SAMPLE_ID, 
                 clones.count = df$CLONOTYPES_TOTAL - 
                   df$CLONOTYPES_FILTERED - ifelse(nc_filter_enabled, df$NON_FUNCTIONAL_CLONOTYPES, 0),
                 migs.count = df$EVENTS_TOTAL - 
                   df$EVENTS_FILTERED - ifelse(nc_filter_enabled, df$NON_FUNCTIONAL_EVENTS, 0),
                 clones.filter.rate = df$CLONOTYPES_FILTERED / df$CLONOTYPES_TOTAL,
                 migs.filter.rate = df$EVENTS_FILTERED / df$EVENTS_TOTAL,
                 clones.nc.rate = df$NON_FUNCTIONAL_CLONOTYPES / df$CLONOTYPES_TOTAL,
                 migs.nc.rate = df$NON_FUNCTIONAL_EVENTS / df$EVENTS_TOTAL
                 )

pal <- colorRampPalette(brewer.pal(9, "GnBu"))(length(unique(df$sample)))
df$sample <-factor(df$sample, levels=df[order(df$clones.count), "sample"])

plotCdrFinal.1 <- function(dd) {
  ggplot(dd, aes(x=sample, fill = sample, y=value)) +
    geom_bar(stat='identity', size = 0.2, colour = "gray30") + coord_flip() +
    geom_text(position = position_dodge(width=1), size = 3, hjust = -0.1, aes(label=value)) +
    scale_y_continuous(expand = c(0, 0), limits=c(0,max(dd$value)*1.2)) +
    scale_fill_manual(values = pal) + guides(fill = F) +
    xlab("") + ylab("") + 
    theme_bw()
}

plotCdrFinal.2 <- function(dd) {
  ggplot(dd, aes(x=sample, fill = sample, y=value)) +
    geom_bar(stat='identity', size = 0.2, colour = "gray30") + 
    geom_text(position = position_dodge(width=1), size = 3, hjust = -0.1, aes(label=round(value*1000)/10)) +
    coord_flip() +    
    scale_fill_manual(values = pal) + guides(fill = F) +
    xlab("") + ylab("") + facet_grid(.~variable) + 
    theme_bw() + scale_y_continuous(labels = percent_format(), expand = c(0, 0), limits=c(0,max(dd$value)*1.2))
}
}
```

### Fig.12 Number of clonotypes found

Below is the plot of sample diversity, i.e. the number of clonotypes in a given sample

```{r, echo=FALSE, message=FALSE, warning=FALSE}
p <- "Nothing to plot"
if (!is.null(cdrfinal_path)) {
df.1 <- data.frame(sample = df$sample, value = df$clones.count)
p<-plotCdrFinal.1(df.1)
}
p
```

### Fig.13 Final number of TCR/Ig molecules

Total number of molecules (MIGs) in final clonotype tables

```{r, echo=FALSE, message=FALSE, warning=FALSE}
p <- "Nothing to plot"
if (!is.null(cdrfinal_path)) {
df.1 <- data.frame(sample = df$sample, value = df$migs.count)
p <- plotCdrFinal.1(df.1)
}
p
```

### Fig.14 Rate of hot-spot and singleton error filtering

Rate of hot-spot and singleton error filtering, in terms of clonotypes (**clone** panel) and MIGs (**mig** panel). Hot-spot errors are indentified by checking whether a given variant is frequently corrected during MIG consensus assembly.

> As clonotypes represented by a single MIG (singletons) have insufficient info to apply MiGEC-style error filtering, a simple frequency-based filtering is used for them.

```{r, echo=FALSE, message=FALSE, warning=FALSE}
p <- "Nothing to plot"
if (!is.null(cdrfinal_path)) {
df.2 <- data.frame(sample = df$sample, mig = df$migs.filter.rate, clone = df$clones.filter.rate)
df.2 <- melt(df.2)
p<-plotCdrFinal.2(df.2)
}
p
```

### Fig.15 Fraction of non-coding clonotypes

Rate of non-coding CDR3 sequences, in terms of clonotypes (**clone** panel) and MIGs (**mig** panel)

```{r, echo=FALSE, message=FALSE, warning=FALSE}
p <- "Nothing to plot"
if (!is.null(cdrfinal_path)) {
df.2 <- data.frame(sample = df$sample, mig = df$migs.nc.rate, clone = df$clones.nc.rate)
df.2 <- melt(df.2)
p<-plotCdrFinal.2(df.2)
}
p
```
